### PR TITLE
Storing map layer selections with localstorage

### LINKED ISSUE #237: Investigate storing Map Layer Selections
### Warframe Hub Issue
---

**Summary (short):** Investigate/implement a solution to store map layer selections so that users don't have to reselect their preferences each time

---
**Description:** The map layer control should have a way to serialize selections. If not, we'll have to do direct manipulations, which I don't like the sound of.

---
**If issue, Reproduction:** N/A

---
**Mockups, screenshots, evidence:** None yet

---

(Check one)
- [ ] Issue
- [x] Suggestion

### PR DESCRIPTION
**Pull request by esunder**
**Summary:**
- Adding localstorage of toggle values for PoeMap and VallisMap. 
- Had to do some refactorings for layers that had differing implementations on how they were added to the leaflet map. e.g. the toroids and caves on the vallis map

---
**Fixes issue (include link):**
closes #237 

https://github.com/WFCD/warframe-hub/issues/237

---
**Mockups, screenshots, evidence:**
Plains:

Vallis:

### CODE REVIEW

**Comment by esunder** [comment]
I fixed the linting issues in the javascript code files. For the json files, I copied the format of the rest of the json files. If I make the suggested fixes, then the new files I created will not follow the same pattern of the old files.

**Comment by EricSihaoLin** [comment]
> I fixed the linting issues in the javascript code files. For the json files, I copied the format of the rest of the json files. If I make the suggested fixes, then the new files I created will not follow the same pattern of the old files.

No problem. I will go ahead and ignore the json files from Codacy.

**Review by EricSihaoLin — COMMENTED** [review]
I have no problem with the new map layer logic.
@TobiTenno Can you comment on the localstorage since you wrote that code?

**Comment by TobiTenno** [comment]
@EricSihaoLin yeah, taking a look

File: src/views/PoEMap.vue [hunk 1]
```diff
@@ -84,116 +80,149 @@ function onEachOddity(feature, layer) {
...
-                  fill: false,
-                })
-                .bindTooltip(feature.properties.name, {
-                  permanent: true,
-                  direction: 'center',
-                  className: 'map-label',
-                })
-                .openTooltip();
-            },
+function dataFn() {
```
Comment (TobiTenno) [inline src/views/PoEMap.vue]: could you give a more meaningful name?

File: src/views/PoEMap.vue [hunk 2]
```diff
@@ -84,116 +80,149 @@ function onEachOddity(feature, layer) {
...
   methods: {
     track() {
       this.$ga.page('/poe/map');
     },
   },
-  mounted: function() {},
+  mounted: function() {
+    this.map = this.$refs.pmap.mapObject;
+    // Get our toggle values from local storage
+    let toggles = localStorage.poeMapToggles ? JSON.parse(localStorage.poeMapToggles) : defaultToggleValues();
```
Comment (TobiTenno) [inline src/views/PoEMap.vue]: could you push them to the state, keeping with how settings are normally loaded?

File: src/views/VallisMap.vue [hunk 3]
```diff
@@ -159,175 +121,192 @@ function onEachOddity(feature, layer) {
...
-        },
-        {
-          name: 'Fishing Spots',
-          json: fishRecommend,
-          opts: {
-            pointToLayer: function(feature, latlng) {
-              return markerAlias(latlng, { icon: fishRecommendMarker });
-            },
-            onEachFeature: onEachFeature,
+function dataFn() {
```
Comment (TobiTenno) [inline src/views/VallisMap.vue]: not a big fan of the naming

File: src/views/VallisMap.vue [hunk 4]
```diff
@@ -159,175 +121,192 @@ function onEachOddity(feature, layer) {
...
+export default {
+  name: 'Vallismap',
+  data: dataFn,
   methods: {
     track() {
       this.$ga.page('/vallis/map');
     },
   },
-  mounted: function() {},
+  mounted: function() {
```
Comment (TobiTenno) [inline src/views/VallisMap.vue]: if we're going to abstract the data function out to a separate one (presumably for space/organization?) could you move mounted to one as well?

**Review by TobiTenno — COMMENTED** [review]
just a few discussion bits, we typically don't store directly to local storage, we route everything through vue's state

it definitely works great, tho, which is 👌 fantastic

**Comment by esunder** [comment]
Thanks for the feedback @TobiTenno.

What do you mean by "we typically don't store directly to local storage, we route everything through vue's state"? 

I am pretty new to Vue, so learning as I go.

File: src/views/VallisMap.vue (hunk 4 above)
Comment (esunder) [reply src/views/VallisMap.vue]: Yep!

File: src/views/VallisMap.vue (hunk 3 above)
Comment (esunder) [reply src/views/VallisMap.vue]: Would getData() be better?

**Comment by TobiTenno** [comment]
oh, all good. there's a system in vue that stores data for configuration and state within the application, stored in a mutable fashion via a store, you can see the guts of the store [here](https://github.com/WFCD/warframe-hub/blob/dev/src/store.js). To append/modify it you just have to write a mutator, and then call the mutator via `this.$store.commit('YOUR_STORE_MUTATOR_NAME', newDatahere)`, and get it from `this.$store.getters.YOUR_STORE_GETTER_NAME`, which typically corresponds to whatever you name the object, at least it does for the ones I do.

Sorry, not trying to make this more complicated for you, just wanna keep all the data access similar

**Comment by esunder** [comment]
I'll take a look, thanks! 

I understand the want for keeping things the same. It keeps the overall code base less complex.

File: src/store.js [hunk 5]
```diff
@@ -34,6 +34,8 @@ const state = {
     xb1: [],
     switch: [],
   },
+  poeMapToggles: {},
```
Comment (TobiTenno) [inline src/store.js]: something small, could we populate them by default with the labels (at least) for the maps so they don't show up blank?

Comment (esunder) [reply src/store.js]: @TobiTenno done!

File: src/views/VallisMap.vue (hunk 3 above)
Comment (TobiTenno) [reply src/views/VallisMap.vue]: either getData or what's there now (`data`) is fine

**Comment by TobiTenno** [comment]
~~@esunder could you rebase your branch? it's my fault cause we had some other changes go in~~

nvm, i'll just push the update branch button

**Review by TobiTenno — APPROVED** [review]
awesome feature, thank you so much.

**Pull request merged**